### ISSUE TITLE
Investigate fast alternative to modulo reduction

### ISSUE THREAD

**Issue by dictoon** [appleseedhq/appleseed#1103]
See
* http://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/
* http://lemire.me/blog/2016/06/30/fast-random-shuffling/

**Comment by lemire** [comment]
I'd be very interested in hearing where you think it might help?

**Comment by dictoon** [comment]
Hi Daniel, thanks for getting in touch.

One place where your technique could be useful is in our set-associative cache implementation:
https://github.com/appleseedhq/appleseed/blob/master/src/appleseed/foundation/utility/cache.h#L767

There are other places where this could be used, such as [the routine that generates integers in a given interval](https://github.com/appleseedhq/appleseed/blob/master/src/appleseed/foundation/math/rng/distribution.h#L119-L136), but they aren't performance-critical.

**Comment by lemire** [comment]
Could this header file help? https://github.com/lemire/fastrange

If you do end up proceeding, I'd be very interested in any benchmark numbers you have. (I'd like to document the benefits... so if it helps appleseed, I'd like to include some numbers and maybe a link.)

**Comment by dictoon** [comment]
Yes, it would help, mostly for properly implementing the 64-bit case.

One thing that wasn't obvious in your article about this trick is that the input value must use the full 32-bit range, otherwise there will be strong bias. In retrospect this is rather obvious, but this will actually prevent us from using the trick in our set-associative cache since we don't have any guarantee about the spread of the hash value.

I'm glad I had a look at your GitHub account, a lot of your work is rather relevant to our needs, especially the fast integer compression codes.

If/when we end up implementing any of this, we'll let you know about our experience and performances.

Thanks again for getting in touch!

**Comment by lemire** [comment]
> One thing that wasn't obvious in your article about this trick is that the input value must use the full 32-bit range, otherwise there will be strong bias

A good hash function should spread hash values over the whole range.

But yes, if you have bad hash functions, then there might be problems.

**Comment by dictoon** [comment]
> A good hash function should spread hash values over the whole range.

Hash values don't necessarily need to cover the full 32-bit range. They could (in the name of performance) be uniformly distributed over a 16-bit subset, for instance. While that would naturally work with a modulo range reduction, it won't with your trick (if I'm not mistaking).

**Comment by lemire** [comment]
@dictoon 

The range computation trick is entirely general... but you do need to know the range of your input. So if it is [0,65536), then you could simply do ``( hash(x) * range ) >> 16``.

I'd be interested in examples where it is cheaper to compute 16-bit hash values than 32-bit hash values.
